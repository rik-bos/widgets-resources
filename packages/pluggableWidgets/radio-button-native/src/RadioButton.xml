<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.native.radiobutton.RadioButton"
        pluginWidget="true"
        needsEntityContext="true"
        offlineCapable="true"
        supportedPlatform="Native"
        xmlns="http://www.mendix.com/widget/1.0/"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Radio Button</name>
    <description>My widget description</description>
    <studioProCategory>Input elements</studioProCategory>
    <studioCategory>Input elements</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="orientation" type="enumeration" defaultValue="horizontal">
                    <caption>Orientation</caption>
                    <description/>
                    <enumerationValues>
                        <enumerationValue key="horizontal">Horizontal</enumerationValue>
                        <enumerationValue key="vertical">Vertical</enumerationValue>
                    </enumerationValues>
                </property>
            </propertyGroup>

            <propertyGroup caption="Data source">
                <property key="enum" type="attribute">
                    <caption>Value</caption>
                    <description/>
                    <attributeTypes>
                        <attributeType name="Enum" />
                    </attributeTypes>
                </property>
            </propertyGroup>

            <propertyGroup caption="Editiability">
                <systemProperty key="Editability"/>
            </propertyGroup>

            <propertyGroup caption="Visibility">
                <systemProperty key="Visibility"/>
            </propertyGroup>

            <propertyGroup caption="Common">
                <systemProperty key="Name"/>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <propertyGroup caption="Events">
                <property key="onChangeAction" type="action" required="false">
                  <caption>On Change</caption>
                  <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
