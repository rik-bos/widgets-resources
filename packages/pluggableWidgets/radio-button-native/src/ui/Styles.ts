import { ViewStyle, TextStyle } from "react-native";
import { Style } from "@mendix/piw-native-utils-internal";

const RADIO_BUTTON_SIZE: number = 20;

export interface RadioButtonStyle extends Style {
    container: ViewStyle;
    radioButtonContainer: ViewStyle;

    caption: TextStyle;
    disabledCaption: TextStyle;

    activeCircleRadioButton: ViewStyle;
    defaultCircleRadioButton: ViewStyle;
    disabledCircleRadioButton: ViewStyle;

    horizontalContainer: ViewStyle;
}

export const defaultRadioButtonStyle: RadioButtonStyle = {
    container: {
        flexDirection: "column",
        flexWrap: "wrap",
        justifyContent: "space-between"
    },
    horizontalContainer: {
        flexDirection: "row"
    },
    radioButtonContainer: {
        flexDirection: "row",
        paddingVertical: 5,
        alignItems: "center"
    },
    caption: {
        marginStart: 5,
        fontSize: 20,
        color: "#000"
    },
    disabledCaption: {
        color: "#CCC"
    },
    defaultCircleRadioButton: {
        backgroundColor: "#000",
        borderColor: "#fff",
        borderWidth: 1,
        height: RADIO_BUTTON_SIZE,
        width: RADIO_BUTTON_SIZE,
        borderRadius: RADIO_BUTTON_SIZE / 2,
        alignItems: "center",
        justifyContent: "center"
    },
    activeCircleRadioButton: {
        backgroundColor: "#fff",
        height: RADIO_BUTTON_SIZE / 2,
        width: RADIO_BUTTON_SIZE / 2,
        borderRadius: RADIO_BUTTON_SIZE / 4
    },
    disabledCircleRadioButton: {
        opacity: 0.7
    }
};
