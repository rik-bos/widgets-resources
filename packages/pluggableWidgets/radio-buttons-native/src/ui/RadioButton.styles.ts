import { Style } from "@mendix/piw-native-utils-internal";
import { ViewStyle, TextStyle } from "react-native";

export const styles = {
    circularBtnStyle: {
        width: 10,
        height: 10,
        borderRadius: 5
    },
    activeBtnStyle: {}
};
