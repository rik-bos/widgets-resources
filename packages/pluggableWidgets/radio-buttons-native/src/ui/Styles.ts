import { Style } from "@mendix/piw-native-utils-internal";
import { TextStyle, ViewStyle } from "react-native";

export interface RadioButtonsStyle extends Style {
    container: ViewStyle;
    containerHorizontal: ViewStyle;
    radioItemContainerStyle: ViewStyle;
    radioItemContainerDisabledStyle: ViewStyle;
    circularBtnStyle: ViewStyle;
    activeBtnStyle: ViewStyle;
    radioItemTitleStyle: TextStyle;
    validationMessage: TextStyle;
}

export const defaultRadioButtonsStyle: RadioButtonsStyle = {
    container: {},
    containerHorizontal: {
        flexDirection: "row",
        alignItems: "center",
        flexWrap: "wrap"
    },
    radioItemContainerStyle: {
        flexDirection: "row",
        alignItems: "center"
    },
    circularBtnStyle: {
        justifyContent: "center",
        alignItems: "center"
    },
    activeBtnStyle: {},
    radioItemTitleStyle: {},
    radioItemTitleDisabledStyle: {},
    circularBtnDisabledStyle: {},
    radioItemContainerDisabledStyle: { opacity: 0.3 },
    validationMessage: {}
};
