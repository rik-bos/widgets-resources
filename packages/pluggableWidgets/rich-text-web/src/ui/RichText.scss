@use "../../../node_modules/ckeditor4/contents.css";
@use "../../../node_modules/ckeditor4/skins/moono-lisa/editor.css";
@use "../../../node_modules/ckeditor4/plugins/tableselection/styles/tableselection.css";
@use "../../../node_modules/ckeditor4/plugins/dialog/styles/dialog.css" as dialog;
@use "../../../node_modules/ckeditor4/plugins/scayt/dialogs/dialog.css" as scaytDialog;
@use "../../../node_modules/ckeditor4/plugins/scayt/skins/moono-lisa/scayt.css";

.widget-rich-text {
    .cke,
    .cke_chrome {
        visibility: visible;
    }

    &.editor-text {
        .cke {
            border: none;
        }

        .cke_top,
        .cke_bottom {
            background: inherit;
            display: none;
        }
    }

    &.editor-bordered {
        .cke {
            border: 1px solid #d1d1d1;
        }

        .cke_top,
        .cke_bottom {
            background: inherit;
            display: none;
        }
    }

    &.editor-borderedToolbar {
    }
}
