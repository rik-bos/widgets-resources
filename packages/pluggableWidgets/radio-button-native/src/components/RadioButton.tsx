import { ReactElement, createElement, useCallback } from "react";
import { View, Text, Pressable } from "react-native";
import { OrientationEnum } from "../../typings/RadioButtonProps";

import { RadioButtonStyle } from "../ui/Styles";

export interface RadioButtonProps {
    caption: string;
    name: string;
    active: boolean;
    orientation: OrientationEnum;
    style: RadioButtonStyle;
    disabled: boolean;
    onRadioButtonPress: (name: string) => void;
}

export function RadioButton({
    caption,
    name,
    active,
    style,
    disabled,
    onRadioButtonPress
}: RadioButtonProps): ReactElement {
    const renderText = useCallback(
        () => <Text style={[style.caption, disabled === true ? style.disabledCaption : null]}>{caption}</Text>,
        [disabled, active, caption]
    );
    const renderRadioCircleButton = useCallback(
        () => (
            <View style={[style.defaultCircleRadioButton, disabled === true ? style.disabledCircleRadioButton : null]}>
                {active ? (
                    <View
                        style={[
                            style.activeCircleRadioButton,
                            disabled === true ? style.disabledCircleRadioButton : null
                        ]}
                    />
                ) : null}
            </View>
        ),
        [disabled, active]
    );

    return (
        <Pressable
            testID={`${name}.${caption}`}
            style={style.radioButtonContainer}
            onPress={() => onRadioButtonPress(name)}
        >
            {renderRadioCircleButton()}
            {renderText()}
        </Pressable>
    );
}
