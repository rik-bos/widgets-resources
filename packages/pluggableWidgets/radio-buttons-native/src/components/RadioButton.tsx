import { createElement, ReactElement } from "react";
import { Pressable, Text, View } from "react-native";
import { RadioButtonsStyle } from "../ui/Styles";

export interface RadioButtonProps {
    title: string;
    active: boolean;
    onSelect: (value: string) => void;
    styles: RadioButtonsStyle;
    name: string;
    disabled: boolean;
}

export function RadioButton({ active, onSelect, title, styles, name, disabled }: RadioButtonProps): ReactElement {
    return (
        <Pressable
            style={[styles.radioItemContainerStyle, disabled && styles.radioItemContainerDisabledStyle]}
            onPress={() => onSelect(name)}
            testID={`radio-button-${name}`}
        >
            <View style={[styles.circularBtnStyle, disabled && styles.circularBtnDisabledStyle]}>
                {active && <View style={styles.activeBtnStyle} />}
            </View>
            <Text style={styles.radioItemTitleStyle}>{title}</Text>
        </Pressable>
    );
}
