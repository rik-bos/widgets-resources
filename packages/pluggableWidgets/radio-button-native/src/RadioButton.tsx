import { createElement, ReactElement, useCallback } from "react";
import { View } from "react-native";
import { flattenStyles } from "@mendix/piw-native-utils-internal";
import { executeAction } from "@mendix/piw-utils-internal";
import { ValueStatus } from "mendix";
import { RadioButton as RadioButtonComponent } from "./components/RadioButton";
import { RadioButtonProps } from "../typings/RadioButtonProps";
import { RadioButtonStyle, defaultRadioButtonStyle } from "./ui/Styles";

export type Props = RadioButtonProps<RadioButtonStyle>;

export function RadioButton(props: RadioButtonProps<RadioButtonStyle>): ReactElement {
    const styles = flattenStyles(defaultRadioButtonStyle, props.style);
    const universe = () => props.enum.universe!;
    const editable = !props.enum.readOnly;

    const onChangeHandler = useCallback(
        pressedItemName => {
            if (pressedItemName === props.enum.value || !editable) {
                return;
            }

            if (props.enum.status === ValueStatus.Available) {
                props.enum.setValue(pressedItemName);
                executeAction(props.onChangeAction);
            }
        },
        [editable, props.enum.value, props.enum.status, props.onChangeAction]
    );

    return (
        <View
            testID={`${props.name}$wrapper`}
            style={[styles.container, props.orientation === "horizontal" ? styles.horizontalContainer : null]}
        >
            {universe()?.map(enumName => (
                <RadioButtonComponent
                    key={`${props.name}.${enumName}`}
                    orientation={props.orientation}
                    style={styles}
                    disabled={props.enum.readOnly}
                    name={enumName}
                    caption={props.enum.formatter.format(enumName)}
                    active={props.enum.value === enumName}
                    onRadioButtonPress={enumName => onChangeHandler(enumName)}
                />
            ))}
        </View>
    );
}
