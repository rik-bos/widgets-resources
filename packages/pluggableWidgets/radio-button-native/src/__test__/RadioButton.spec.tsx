import { createElement } from "react";
import { Text } from "react-native";
import { EditableValueBuilder, actionValue } from "@mendix/piw-utils-internal";
import { render, fireEvent } from "@testing-library/react-native";

import { RadioButton, Props } from "../RadioButton";
import { defaultRadioButtonStyle } from "../ui/Styles";

const name = "radioButton1";
const createProps = (props?: Partial<Props>): Props => {
    const style = props?.style ?? {};
    const defaultProps: Props = {
        name,
        style: [{ ...defaultRadioButtonStyle, ...style }],
        orientation: "vertical",
        enum: new EditableValueBuilder<string>().withUniverse("a", "b").withValue("a").build()
    };

    return { ...defaultProps, ...props };
};

describe("RadioButton", () => {
    it("renders with default styles", () => {
        const props = createProps();
        const component = render(<RadioButton {...props} />);
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders captions", () => {
        const props = createProps();
        const component = render(<RadioButton {...props} />);

        const textComponents = component.UNSAFE_getAllByType(Text);
        expect(textComponents).toHaveLength(2);
        expect(textComponents[0].props.children).toEqual("Formatted a");
        expect(textComponents[1].props.children).toEqual("Formatted b");
    });

    it("renders vertical", () => {
        const props = createProps({
            orientation: "vertical"
        });

        const component = render(<RadioButton {...props} />);

        expect(component.getByTestId(`${name}$wrapper`).props.style).toEqual(
            expect.arrayContaining([expect.objectContaining({ flexDirection: "column" })])
        );
    });

    it("renders horizontal", () => {
        const props = createProps({
            orientation: "horizontal"
        });

        const component = render(<RadioButton {...props} />);

        expect(component.getByTestId(`${name}$wrapper`).props.style).toEqual(
            expect.arrayContaining([expect.objectContaining({ flexDirection: "row" })])
        );
    });

    it("ignores pressing the active value", () => {
        const props = createProps({
            onChangeAction: actionValue()
        });
        const component = render(<RadioButton {...props} />);

        fireEvent.press(component.getByText("Formatted a"));

        expect(props.enum.setValue).not.toHaveBeenCalled();
    });

    it("ignores pressing when non editable", () => {
        const props = createProps({
            enum: new EditableValueBuilder<string>().withUniverse("a", "b").withValue("a").isReadOnly().build()
        });

        const component = render(<RadioButton {...props} />);

        fireEvent.press(component.getByText("Formatted b"));

        expect(props.enum.setValue).not.toHaveBeenCalled();
    });

    it("ignores pressing when not available", () => {
        const props = createProps({
            enum: new EditableValueBuilder<string>().withUniverse("a", "b").withValue("a").isUnavailable().build()
        });

        const component = render(<RadioButton {...props} />);

        fireEvent.press(component.getByText("Formatted b"));

        expect(props.enum.setValue).not.toHaveBeenCalled();
    });

    it("sets the value when pressed and executes the on change action", () => {
        const props = createProps({
            onChangeAction: actionValue()
        });

        const component = render(<RadioButton {...props} />);

        fireEvent.press(component.getByText("Formatted b"));

        expect(props.enum.setValue).toHaveBeenCalledWith("b");
        expect(props.onChangeAction?.execute).toHaveBeenCalledTimes(1);
    });
});
