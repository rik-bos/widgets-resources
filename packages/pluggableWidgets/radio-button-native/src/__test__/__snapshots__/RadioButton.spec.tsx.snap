// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton renders with default styles 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "column",
        "flexWrap": "wrap",
        "justifyContent": "space-between",
      },
      null,
    ]
  }
  testID="radioButton1$wrapper"
>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingVertical": 5,
      }
    }
    testID="a.Formatted a"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#000",
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 20,
            "justifyContent": "center",
            "width": 20,
          },
          null,
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
              "borderRadius": 5,
              "height": 10,
              "width": 10,
            },
            null,
          ]
        }
      />
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "marginStart": 5,
          },
          null,
        ]
      }
    >
      Formatted a
    </Text>
  </View>
  <View
    accessible={true}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "paddingVertical": 5,
      }
    }
    testID="b.Formatted b"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#000",
            "borderColor": "#fff",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 20,
            "justifyContent": "center",
            "width": 20,
          },
          null,
        ]
      }
    />
    <Text
      style={
        Array [
          Object {
            "color": "#000",
            "fontSize": 20,
            "marginStart": 5,
          },
          null,
        ]
      }
    >
      Formatted b
    </Text>
  </View>
</View>
`;
